IBX-2532: Added new Address field type (#1721)
ezsystems/developer-documentation: docs/api/field_types_reference/addressfield.md

# Address Field Type

This Field represents and handles address fields.

> It allows you to customize address Field Types per country.

Reviewer: ```suggestion
It allows you to customize address Fields per country.
```

| Name      | Internal name   | Expected input              |
|-----------|-----------------|-----------------------------|
| `Address` | `ibexa_address` | `string`, `string`, `array` |

The Address Field Type is available via the Address Bundle

> provided by the [ibexa/fieldtype-address](https://github.com/ibexa/fieldtype-address) package.

Reviewer: ```suggestion
provided by the `ibexa/fieldtype-address` package.
```
The repo is private, so link would lead nowhere for most readers

## PHP API Field Type

### Inputs:

| Type     | Description                                   | Example           |
|----------|-----------------------------------------------|-------------------|
| `string` | Name of the address.                          | `My home address` |
| `string` | Country code in ISO 3166-1 alpha-2 format.    | `PL`              |
| `array`  | Additional fields, defined by address format. | see below         |

> ### Example input:

Reviewer: ```suggestion
### Example input
```

```php
new FieldType\Value(
    'My home address',
    'PL',
    [
        'city' => 'Warsaw',
        'region' => 'Masovian',
        'postal_code' => '11-123',
    ]
);
```

### Validation

> This Field Type validates whether `Country` and `Name` fields have been field out.

Reviewer: ```suggestion
This Field Type validates whether `Country` and `Name` fields have been filled out.
```

> ### Value Object:

Reviewer: ```suggestion
### Value Object
```

> #### Properties:

Reviewer: ```suggestion
#### Properties
```

| Property   | Type     | Description                                   |
|------------|----------|-----------------------------------------------|
| `$name`    | `string` | Name of the address.                          |
| `$country` | `string` | Country code in ISO 3166-1 alpha-2 format.    |
| `$fields`  | `array`  | Additional fields, defined by address format. |

> #### Constructor:

Reviewer: ```suggestion
#### Constructor
```

See above (Example input).

> ### Formats:

Reviewer: ```suggestion
### Formats
```

> Default configuration.

Reviewer: ```suggestion
The following default configuration defines default fields for `personal` address type:
```

```yaml
formats:
    personal:
        country:
            default:
                - region
                - locality
                - street
                - postal_code
```

> 
> Defines default fields for `personal` address type.

Reviewer: ```suggestion
```

> #### Extending:

Reviewer: ```suggestion
#### Extending
```
or maybe:
```suggestion
#### Modifying Field configuration
```

```yaml
formats:
    billing_address:
        country:
            DE:
                - tax_number
                - city
                - address
                - postal_code
```

Adds (or alters) an address format for `DE` country of `billing_address` type.

> ### Field form types:

Reviewer: ```suggestion
### Field form types
```

By default, each field is a simple text input with a label made of field identifier.
To change the type of field, you need to listen to a specific event.

> For each field below events are emitted (in order):

Reviewer: ```suggestion
For each field below events are dispatched (in order):
```
emit is fine, but we use dispatch elsewhere (e.g. https://doc.ibexa.co/en/master/guide/repository/event_reference/event_reference/)

```
ibexa.address.field.{FIELD_IDENTIFIER}
ibexa.address.field.{FIELD_IDENTIFIER}.{ADDRESS_TYPE}
ibexa.address.field.{FIELD_IDENTIFIER}.{ADDRESS_TYPE}.{COUNTRY_CODE}
```

> #### Example:

Reviewer: ```suggestion
#### Example
```

```
ibexa.address.field.tax_number
ibexa.address.field.tax_number.billing_address
ibexa.address.field.tax_number.billing_address.DE
```

> #### Example event listener:

Reviewer: ```suggestion
#### Example event listener
```

```php
use Ibexa\Contracts\FieldTypeAddress\Event\MapFieldEvent;
use Symfony\Component\EventDispatcher\EventSubscriberInterface;
use Symfony\Component\Form\Extension\Core\Type\IntegerType;

> class ExampleAddressSubscriber implements EventSubscriberInterface

Reviewer: ```suggestion

class ExampleAddressSubscriber implements EventSubscriberInterface
```
I think, but cs fixer should confirm
